articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+A managed life cycle for nodes allows greater precision over the state of ROS system. It will allow roslaunch to have greater control over launching a ROS system, and in particular will allow roslaunch to ensure that all components have been instantiated correctly before it allows any component to begin executing its behaviour. It will also allow nodes to be restarted or replaced on-line.
+
+The most important concept of this document is that a managed node presents a known interface, executes according to a known life cycle state machine, and otherwise can be considered a black box. This allows freedom to the node developer on how they provide the managed life cycle functionality, while also ensuring that any tools created for managing nodes can work with any compliant node.
+
+
+## Glossary
+
+* Configuration: To alter aspects of a node's behaviour through known interfaces (services and topics). This includes setting parameters and connecting topics. Configuration may be performed multiple times during a node's life.
+* Initialise: To shift a node from the created state to the inactive state, during which the node is prepared for execution. For example, memory buffers can be allocated.
+* Activate: To shift a node from the inactive state to the active state. For example, internal variables can be re-initialised based on changed configuration parameters and hardware access may be obtained.
+* Deactivate: To shift a node from the active state to the inactive state. For example, hardware access may be relinquished.
+* Fatal error: An unforeseen error that cannot be handled by the robot application, requiring the node to be reset or destroyed.
+
+
+## Life cycle
+
+![The proposed node life cycle state machine](/img/node_lifecycle/life_cycle_sm.png "The proposed node life cycle state machine")
+
+The behaviour of each state is as defined below.
+
+### Created state
+
+This is the life cycle state the node is in immediately after being instantiated.
+
+The node may transition to the `inactive` state by being initialised. Initialisation of a node will typically involve those tasks that must be performed once during the node's life time, such as obtaining permanent memory buffers and setting up topic publications/subscriptions that do not change.
+
+Implementation note: In object-oriented languages, the onCreated method can be mapped to the class constructor.
+
+### Inactive state
+
+This state represents a node that is not currently performing any processing.
+
+The main purpose of this state is to allow a node to be (re-)configured (changing configuration parameters, adding and removing topic publications/subscriptions, etc.) without altering its behaviour while it is running.
+
+[To discuss: Should there be a callback for when configuration is changed, e.g. altered parameters, new topics, etc.?]
```

Comment:
I also think that _when_ the parameters are read is a key issue tied with the managed lifecycle.

Delegating _when_ to read the parameters to the node developers, such as in ROS 1, generates too much ambiguity in my opinion. Therefore, I lean towards defining different types of parameters, with varying degrees of reading restrictions:

- read on initialisation (parameters which affect lengthy/blocking configuration actions which would therefore be performed on `onInitialised`, e.g. size of buffers to allocate, files to open)
- read on activation/start (everything not involved in lengthy configuration operations but which must not be modified while the node is active)
- read continuously (`dynamic_reconfigure` style, although whether such a parameter is semantically different from a subscription could be up for discussion)

This would allow much better understanding of a node's behavior from it's specification, which I think is a great benefit, while not adding too much complexity. Thoughts?